### PR TITLE
Implement most trivial members.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/libsolidity/semanticTests/viaYul/comparison.sol [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    function gte(uint a, uint b) public pure returns (bool) {
+        return a >= b;
+    }
+    function sgte(int a, int b) public pure returns (bool) {
+        return a >= b;
+    }
+    function eq(uint a, uint b) public pure returns (bool) {
+        return a == b;
+    }
+    function neq(uint a, uint b) public pure returns (bool) {
```
Comment (Marenz) [inline test/libsolidity/semanticTests/viaYul/comparison.sol]: What about tests for differently sized (u)ints, 8,16,32, etc?

Comment (leonardoalt) [reply test/libsolidity/semanticTests/viaYul/comparison.sol]: I think the type conversion PR tested those.

File: test/libsolidity/semanticTests/viaYul/comparison.sol [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+// lte(uint256,uint256): 6, 5 -> false
+// gte(uint256,uint256): 4, 5 -> false
+// gte(uint256,uint256): 5, 5 -> true
+// gte(uint256,uint256): 6, 5 -> true
+// eq(uint256,uint256): 4, 5 -> false
+// eq(uint256,uint256): 5, 5 -> true
+// eq(uint256,uint256): 6, 5 -> false
+// ne(uint256,uint256): 4, 5 -> false
+// ne(uint256,uint256): 5, 5 -> true
+// ne(uint256,uint256): 6, 5 -> false
```
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/comparison.sol]: The 3 above should be `neq`. Also the results are inverted.

File: test/libsolidity/semanticTests/viaYul/msg_sender.sol [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    function test() public view returns (bool) {
+        address x;
+        assembly { x := caller() }
+        return x == msg.sender;
+    }
+}
+// ====
+// compileViaYul: true
+// ----
+// test() -> 1
```
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/msg_sender.sol]:
```suggestion
// test() -> true
```

File: test/libsolidity/semanticTests/viaYul/comparison.sol [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+// lte(uint256,uint256): 4, 5 -> true
+// lte(uint256,uint256): 5, 5 -> true
+// lte(uint256,uint256): 6, 5 -> false
+// gte(uint256,uint256): 4, 5 -> false
+// gte(uint256,uint256): 5, 5 -> true
+// gte(uint256,uint256): 6, 5 -> true
+// eq(uint256,uint256): 4, 5 -> false
+// eq(uint256,uint256): 5, 5 -> true
+// eq(uint256,uint256): 6, 5 -> false
+// ne(uint256,uint256): 4, 5 -> false
```
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/comparison.sol]:
```suggestion
// neq(uint256,uint256): 4, 5 -> true
```

Comment (leonardoalt) [reply test/libsolidity/semanticTests/viaYul/comparison.sol]: Also the two below

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 5]
```diff
@@ -314,6 +355,154 @@ bool IRGeneratorForStatements::visit(FunctionCall const& _functionCall)
...
+		if (member == "balance")
+			defineExpression(_memberAccess) <<
+				"balance(" <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::address()) <<
+				")\n";
+		else if ((set<string>{"send", "transfer"}).count(member))
+		{
+			solAssert(dynamic_cast<AddressType const&>(*_memberAccess.expression().annotation().type).stateMutability() == StateMutability::Payable, "");
+			defineExpression(_memberAccess) <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::payableAddress()) <<
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:448]: Is this incomplete?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 6]
```diff
@@ -314,6 +355,154 @@ bool IRGeneratorForStatements::visit(FunctionCall const& _functionCall)
...
+				"\n";
+		}
+		else if ((set<string>{"call", "callcode", "delegatecall", "staticcall"}).count(member))
+			defineExpression(_memberAccess) <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::address()) <<
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:453]: Is this incomplete?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 7]
```diff
@@ -314,6 +355,154 @@ bool IRGeneratorForStatements::visit(FunctionCall const& _functionCall)
...
+			solAssert(false, "Blockhash has been removed.");
+		else if (member == "creationCode" || member == "runtimeCode")
+		{
+			solUnimplementedAssert(false, "");
+		}
+		else if (member == "name")
+		{
+			solUnimplementedAssert(false, "");
+		}
+		else if ((set<string>{"encode", "encodePacked", "encodeWithSelector", "encodeWithSignature", "decode"}).count(member))
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		else if (set<string>{"encode", "encodePacked", "encodeWithSelector", "encodeWithSignature", "decode"}.count(member))
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 8]
```diff
@@ -314,6 +355,154 @@ bool IRGeneratorForStatements::visit(FunctionCall const& _functionCall)
...
+		break;
+	}
+	case Type::Category::Address:
+	{
+		if (member == "balance")
+			defineExpression(_memberAccess) <<
+				"balance(" <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::address()) <<
+				")\n";
+		else if ((set<string>{"send", "transfer"}).count(member))
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		else if (set<string>{"send", "transfer"}.count(member))
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 9]
```diff
@@ -314,6 +355,154 @@ bool IRGeneratorForStatements::visit(FunctionCall const& _functionCall)
...
+		{
+			solAssert(dynamic_cast<AddressType const&>(*_memberAccess.expression().annotation().type).stateMutability() == StateMutability::Payable, "");
+			defineExpression(_memberAccess) <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::payableAddress()) <<
+				"\n";
+		}
+		else if ((set<string>{"call", "callcode", "delegatecall", "staticcall"}).count(member))
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		else if (set<string>{"call", "callcode", "delegatecall", "staticcall"}.count(member))
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 10]
```diff
@@ -314,6 +355,154 @@ bool IRGeneratorForStatements::visit(FunctionCall const& _functionCall)
...
+		else
+			solAssert(false, "Unknown magic member.");
+		break;
+	case Type::Category::Struct:
+	{
+		solUnimplementedAssert(false, "");
+	}
+	case Type::Category::Enum:
+	{
+		EnumType const& type = dynamic_cast<EnumType const&>(*_memberAccess.expression().annotation().type);
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:522]: Is this cast always true?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 11]
```diff
@@ -330,7 +519,25 @@ bool IRGeneratorForStatements::visit(Identifier const& _identifier)
...
+		switch (magicVar->type()->category())
+		{
+		case Type::Category::Contract:
+			// "this" or "super"
+			if (!dynamic_cast<ContractType const&>(*magicVar->type()).isSuper())
+				value = "address()";
+			else
+				return false;
+		case Type::Category::Integer:
+			// "now"
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Might want to assert? :p

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 12]
```diff
@@ -330,7 +519,25 @@ bool IRGeneratorForStatements::visit(Identifier const& _identifier)
...
+	if (MagicVariableDeclaration const* magicVar = dynamic_cast<MagicVariableDeclaration const*>(declaration))
+	{
+		switch (magicVar->type()->category())
+		{
+		case Type::Category::Contract:
+			// "this" or "super"
+			if (!dynamic_cast<ContractType const&>(*magicVar->type()).isSuper())
+				value = "address()";
+			else
+				return false;
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Why false?

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Please rebase and add tests for the members.

File: test/libsolidity/semanticTests/viaYul/comparison.sol (hunk 1 above)
Comment (leonardoalt) [reply test/libsolidity/semanticTests/viaYul/comparison.sol]: It didn't.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 5 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:448]: It's fine - it is just the function `x.send`, not the function call `x.send()`.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 6 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:453]: Same here, just the function.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 10 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:522]: If not, it will throw...

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:522]: This code was adapted from the classic code generator, so I would say it is fine.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**